In [ ]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
path_abalone_data= '/content/drive/MyDrive/Colab Notebooks/Machine Learning 1/Datasets Machine Learning 1/abalone.csv'

In [ ]:
abalone_df= pd.read_csv(path_abalone_data)

In [ ]:
abalone_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Sex             4177 non-null   object 
 1   Length          4177 non-null   float64
 2   Diameter        4177 non-null   float64
 3   Height          4177 non-null   float64
 4   Whole weight    4177 non-null   float64
 5   Shucked weight  4177 non-null   float64
 6   Viscera weight  4177 non-null   float64
 7   Shell weight    4177 non-null   float64
 8   Rings           4177 non-null   int64  
dtypes: float64(7), int64(1), object(1)
memory usage: 293.8+ KB


In [ ]:
abalone_df.rename(columns={"Whole weight": "Whole_weight", "Shucked weight": "Shucked_weight","Viscera weight":"Viscera_weight", "Shell weight":"Shell_weight" }, inplace= True)

In [ ]:
abalone_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Sex             4177 non-null   object 
 1   Length          4177 non-null   float64
 2   Diameter        4177 non-null   float64
 3   Height          4177 non-null   float64
 4   Whole_weight    4177 non-null   float64
 5   Shucked_weight  4177 non-null   float64
 6   Viscera_weight  4177 non-null   float64
 7   Shell_weight    4177 non-null   float64
 8   Rings           4177 non-null   int64  
dtypes: float64(7), int64(1), object(1)
memory usage: 293.8+ KB


# Question 1

#1a

In [ ]:
abalone_df_without_sex = abalone_df.drop('Sex', axis=1)

abalone_df_without_sex.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Length          4177 non-null   float64
 1   Diameter        4177 non-null   float64
 2   Height          4177 non-null   float64
 3   Whole_weight    4177 non-null   float64
 4   Shucked_weight  4177 non-null   float64
 5   Viscera_weight  4177 non-null   float64
 6   Shell_weight    4177 non-null   float64
 7   Rings           4177 non-null   int64  
dtypes: float64(7), int64(1)
memory usage: 261.2 KB


In [ ]:
abalone_df_without_sex.iloc[:,:-1]

Length  Diameter  Height  Whole_weight  Shucked_weight  Viscera_weight  \
0      0.455     0.365   0.095        0.5140          0.2245          0.1010   
1      0.350     0.265   0.090        0.2255          0.0995          0.0485   
2      0.530     0.420   0.135        0.6770          0.2565          0.1415   
3      0.440     0.365   0.125        0.5160          0.2155          0.1140   
4      0.330     0.255   0.080        0.2050          0.0895          0.0395   
...      ...       ...     ...           ...             ...             ...   
4172   0.565     0.450   0.165        0.8870          0.3700          0.2390   
4173   0.590     0.440   0.135        0.9660          0.4390          0.2145   
4174   0.600     0.475   0.205        1.1760          0.5255          0.2875   
4175   0.625     0.485   0.150        1.0945          0.5310          0.2610   
4176   0.710     0.555   0.195        1.9485          0.9455          0.3765   

      Shell_weight  
0           0.1500  
1           0.0700  
2           0.2100  
3           0.1550  
4           0.0550  
...            ...  
4172        0.2490  
4173        0.2605  
4174        0.3080  
4175        0.2960  
4176        0.4950  

[4177 rows x 7 columns]

In [ ]:
X_all= abalone_df_without_sex.iloc[:,:-1]

In [ ]:
y= abalone_df_without_sex['Rings']

In [ ]:
X_train, X_test, y_train, y_test= train_test_split (X_all, y, test_size=0.2, random_state= 1)

In [ ]:
%run "/content/drive/MyDrive/Colab Notebooks/Machine Learning 1/Module 2/BSS_definitions.ipynb"

In [ ]:
# Invoke the 'my_best_subset_selection_cv()' and store results in a data frame ( I called the data frame 'bss_cv_results')

bss_cv_results= my_best_subset_selection_cv ( X_train , y_train, folds= 10)

<ipython-input-13-4e98b24c1dab>:40: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  out_df = pd.concat([out_df, df_iteration], ignore_index=True)


Cross-validation results for predictors: ['Shell_weight']
Cross-validation MSE values: [5.9049175  6.00498261 5.11222711 6.23101732 5.98309867 8.02097938
 6.04962547 5.5825228  6.28083269 7.8741305 ]
Mean cross-validation MSE: 6.3044

Cross-validation results for predictors: ['Shucked_weight', 'Shell_weight']
Cross-validation MSE values: [4.78350001 5.52468573 4.62878317 5.59301829 5.38469515 6.95206122
 5.22029299 4.71163047 5.30696182 6.66518952]
Mean cross-validation MSE: 5.4771

Cross-validation results for predictors: ['Diameter', 'Shucked_weight', 'Shell_weight']
Cross-validation MSE values: [4.55014933 4.97418946 4.23539704 5.18619663 5.11438686 6.7301937
 4.9642418  4.61216304 5.03001405 6.31874828]
Mean cross-validation MSE: 5.1716

Cross-validation results for predictors: ['Diameter', 'Whole_weight', 'Shucked_weight', 'Shell_weight']
Cross-validation MSE values: [4.36740009 5.02667326 4.23351929 4.89634926 4.89658057 6.63986297
 4.76000322 4.46365202 4.91915367 6.27805055]
Me

In [ ]:
# Show content stored in bss_cv_results

pd.set_option('display.max_colwidth', None)

bss_cv_results

Predictors  \
0                                                                          [Shell_weight]   
1                                                          [Shucked_weight, Shell_weight]   
2                                                [Diameter, Shucked_weight, Shell_weight]   
3                                  [Diameter, Whole_weight, Shucked_weight, Shell_weight]   
4                          [Diameter, Height, Whole_weight, Shucked_weight, Shell_weight]   
5          [Diameter, Height, Whole_weight, Shucked_weight, Viscera_weight, Shell_weight]   
6  [Length, Diameter, Height, Whole_weight, Shucked_weight, Viscera_weight, Shell_weight]   

   Mean_CV_MSE  
0       6.3044  
1       5.4771  
2       5.1716  
3       5.0481  
4       5.0961  
5       5.0675  
6       5.0739

In [ ]:
# Set up threshold for minimun acceptable % reduction in Mean CV MSE

percent_reduction_threshold= 10

In [ ]:
# Compute percent reduction in Mean_CV_MSE from one model to the next one
percent_reduction = - ( bss_cv_results['Mean_CV_MSE'].pct_change().dropna() )*100
percent_reduction.name = "Pct reduc from previous model"

# Find the last set of predictors that meet the threshold
last_row_above_threshold = bss_cv_results.iloc[percent_reduction[percent_reduction >= percent_reduction_threshold].index[-1]]['Predictors']

print("BSS selects the following predictors for the model:")
print(last_row_above_threshold)
print ()
print("Adding additional predictors results in a % reduction in Mean CV MSE below", percent_reduction_threshold, "%:")

BSS selects the following predictors for the model:
['Shucked_weight', 'Shell_weight']

Adding additional predictors results in a % reduction in Mean CV MSE below 10 %:


In [ ]:
regression_final_training= LinearRegression()

In [ ]:
regression_final_training.fit (X_train[['Shucked_weight', 'Shell_weight']], y_train)

LinearRegression()

In [ ]:
y_predicted_final_test =regression_final_training.predict(X_test[['Shucked_weight', 'Shell_weight']])

In [ ]:
# Extract coefficients
intercept = regression_final_training.intercept_
coefficients = regression_final_training.coef_

# Print the model equation
print("The model equation is:")
print(f"Rings = {intercept:.4f} + ({coefficients[0]:.4f} * Shucked_weight) + ({coefficients[1]:.4f} * Shell_weight)")

The model equation is:
Rings = 6.6363 + (-8.7368 * Shucked_weight) + (26.8665 * Shell_weight)


#1b

In [ ]:
rmse = mean_squared_error(y_test, y_predicted_final_test, squared= False)

print("Root Mean Squared Error on the test set is:", rmse)

Root Mean Squared Error on the test set is: 2.34212412984446


In [ ]:
# Evaluate the RMSE
mean_rings = y_train.mean()
std_rings = y_train.std()

print("Mean of Rings in the training set:", mean_rings)
print("Standard Deviation of Rings in the training set:", std_rings)
print("RMSE as a percentage of the mean value:", (rmse / mean_rings) * 100)

Mean of Rings in the training set: 9.932954205327746
Standard Deviation of Rings in the training set: 3.2475223970220286
RMSE as a percentage of the mean value: 23.579330795547342


# Justification for Question 1

## The RMSE being about 23.58% of the mean indicates that the model's predictions deviate by nearly a quarter from the average Rings value, which is quite significant. This high RMSE suggests that the model's errors are substantial relative to what it aims to predict.

## Moreover, with an RMSE of 2.34, nearly 72% of the standard deviation (3.25), it's evident that the model may not be effectively capturing the data's variability. This performance is not much better than a simple model that would predict the average for all observations, highlighting a need for potential improvements in model complexity or methodology.

# Question 2

# 2a

In [ ]:
# Correlation matrix for 'Rings' with all other numerical columns

correlation_matrix = abalone_df_without_sex.corr()
rings_correlation = correlation_matrix['Rings']
print(rings_correlation)

Length            0.556720
Diameter          0.574660
Height            0.557467
Whole_weight      0.540390
Shucked_weight    0.420884
Viscera_weight    0.503819
Shell_weight      0.627574
Rings             1.000000
Name: Rings, dtype: float64


# Justification

## The predictor Shell_weight is selected for the simple linear regression analysis to predict Rings because it has the highest correlation value of 0.627574, indicating the strongest linear relationship with the target variable.

# 2b

In [ ]:
regression_model_Shell_weight = LinearRegression()

In [ ]:
X_Shell_weight = np.array(abalone_df_without_sex['Shell_weight']).reshape(-1,1)

In [ ]:
regression_model_Shell_weight.fit ( X_Shell_weight , y)

LinearRegression()

In [ ]:
y_predicted_simple_reg= regression_model_Shell_weight.predict(X_Shell_weight)

In [ ]:
r2_value = r2_score(y, y_predicted_simple_reg)

r2_value

0.39384918134304336

In [ ]:
# Adjusted R squared

adj_r2_linear = 1 - (1- r2_value) * ( (y.size - 1) / (y.size - 1 - 1) )

adj_r2_linear

0.39370399551821533

In [ ]:
from sklearn.preprocessing import PolynomialFeatures

In [ ]:
# 'include_bias' means whether you want to include the intercept in the equation.
# No need to include intercept now since the function LinearRegression () will include by default.

poly2_object= PolynomialFeatures(degree=2, include_bias=False)

In [ ]:
X_Shell_weight_poly2 = poly2_object.fit_transform(X_Shell_weight)
X_Shell_weight_poly2

array([[0.15    , 0.0225  ],
       [0.07    , 0.0049  ],
       [0.21    , 0.0441  ],
       ...,
       [0.308   , 0.094864],
       [0.296   , 0.087616],
       [0.495   , 0.245025]])

In [ ]:
model_poly2_Shell_weight = LinearRegression()

In [ ]:
model_poly2_Shell_weight.fit(X_Shell_weight_poly2, y)

LinearRegression()

In [ ]:
y_pred_poly2= model_poly2_Shell_weight.predict(X_Shell_weight_poly2)

In [ ]:
r2_value_poly2= r2_score(y, y_pred_poly2)

In [ ]:
# Adjusted R squared

1 - (1- r2_value_poly2) * ( (y.size - 1) / (y.size - 2 - 1) ) # The 2 represents that we have two predictors in the model: X and X squared

0.4080939406427013

In [ ]:
adjusted_r2_linear = 0.39370399551821533
adjusted_r2_poly = 0.4080939406427013

percent_improvement = ((adjusted_r2_poly - adjusted_r2_linear) / adjusted_r2_linear) * 100
print(f"Percentage Improvement: {percent_improvement:.4f}%")


Percentage Improvement: 3.6550%


In [ ]:
intercept = model_poly2_Shell_weight.intercept_
coefficients = model_poly2_Shell_weight.coef_

# Assign coefficients to the terms
beta_0 = intercept
beta_1 = coefficients[0]
beta_2 = coefficients[1]

# Print the formula
print(f"Rings = {beta_0:.4f} + ({beta_1:.4f} * Shell_weight) + ({beta_2:.4f} * Shell_weight^2)")

Rings = 5.6636 + (22.2613 * Shell_weight) + (-13.6965 * Shell_weight^2)


# Justification

##Since the improvement in Adjusted R-squared is above the 3% threshold, adding the squared term of the predictor (Shell_weight) is justified. This means that the polynomial model (including the squared term) provides a statistically significant improvement over the simple linear model.

# 2c

In [ ]:
predictor_list_2= abalone_df_without_sex.columns.difference(['Shell_weight','Rings'])
predictor_list_2

Index(['Diameter', 'Height', 'Length', 'Shucked_weight', 'Viscera_weight',
       'Whole_weight'],
      dtype='object')

In [ ]:
X_all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Length          4177 non-null   float64
 1   Diameter        4177 non-null   float64
 2   Height          4177 non-null   float64
 3   Whole_weight    4177 non-null   float64
 4   Shucked_weight  4177 non-null   float64
 5   Viscera_weight  4177 non-null   float64
 6   Shell_weight    4177 non-null   float64
dtypes: float64(7)
memory usage: 228.6 KB


In [ ]:
# Polynomial features for 'horsepower'
poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly_predictor = poly.fit_transform(X_all[['Shell_weight']])

# List to store adjusted R-squared values for each combination of poly-horsepower and predictor
result_list = []

# Loop through each predictor
for loop_predictor in predictor_list_2:

    # Add the current predictor to the feature matrix
    X_poly_plus_x = np.column_stack((X_poly_predictor , X_all[[loop_predictor]]))

    # Fit a linear regression model
    model = LinearRegression()
    model.fit(X_poly_plus_x, y)

    # Calculate adjusted R-squared
    y_pred = model.predict(X_poly_plus_x)
    r2_value_loop= r2_score (y, y_pred)
    adj_r_sq_model= np.round(1 - (1- r2_value_loop) * ( (y.size - 1) / (y.size - 3 - 1) ), 4) # 3 bc we have 3 predictors: hpw, hpw sq, and the iteration predictor

    # Append the results to the list
    result_list.append({"Predictor_added_to_poly": loop_predictor, "Adj_R_Squared": adj_r_sq_model})

# Create a data frame from the list of dictionaries
adding_to_poly_df = pd.concat([pd.DataFrame(item, index=[0]) for item in result_list], ignore_index=True)

adding_to_poly_df =adding_to_poly_df.sort_values(by='Adj_R_Squared', ascending=False)

In [ ]:
# Output data frame

adding_to_poly_df

Predictor_added_to_poly  Adj_R_Squared
3          Shucked_weight         0.5120
5            Whole_weight         0.4653
4          Viscera_weight         0.4477
2                  Length         0.4286
0                Diameter         0.4191
1                  Height         0.4092

In [ ]:
adding_to_poly_df[ adding_to_poly_df['Adj_R_Squared']== max(adding_to_poly_df['Adj_R_Squared']) ]

Predictor_added_to_poly  Adj_R_Squared
3          Shucked_weight          0.512

In [ ]:
(0.512 - 0.4080939406427013)/0.4080939406427013

0.2546130902940105

# Justification

##The highest Adjusted R-squared is obtained for the model that combines the second-degree polynomial of Shell_weight with the linear term of Shucked_weight. Adding the linear term of Shucked_weight to the polynomial model based on Shell_weight produces a significant increase in Adjusted R-squared. Adjusted R-squared increases from 40.8% to 51.2% (a 25.46% increase). This substantial improvement justifies the inclusion of Shucked_weight in the model, as it contributes meaningfully to the model’s explanatory power.

# Question 3

#3a

In [ ]:
abalone_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Sex             4177 non-null   object 
 1   Length          4177 non-null   float64
 2   Diameter        4177 non-null   float64
 3   Height          4177 non-null   float64
 4   Whole_weight    4177 non-null   float64
 5   Shucked_weight  4177 non-null   float64
 6   Viscera_weight  4177 non-null   float64
 7   Shell_weight    4177 non-null   float64
 8   Rings           4177 non-null   int64  
dtypes: float64(7), int64(1), object(1)
memory usage: 293.8+ KB


In [ ]:
abalone_df['Sex'].value_counts()

Sex
M    1528
I    1342
F    1307
Name: count, dtype: int64

In [ ]:
abalone_df_dummies = pd.get_dummies(abalone_df,columns= ['Sex'], drop_first= True)

In [ ]:
abalone_df_dummies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Length          4177 non-null   float64
 1   Diameter        4177 non-null   float64
 2   Height          4177 non-null   float64
 3   Whole_weight    4177 non-null   float64
 4   Shucked_weight  4177 non-null   float64
 5   Viscera_weight  4177 non-null   float64
 6   Shell_weight    4177 non-null   float64
 7   Rings           4177 non-null   int64  
 8   Sex_I           4177 non-null   bool   
 9   Sex_M           4177 non-null   bool   
dtypes: bool(2), float64(7), int64(1)
memory usage: 269.3 KB


In [ ]:
X_Shell_weight_origin = abalone_df_dummies[['Shell_weight', 'Sex_I', 'Sex_M']].copy()

In [ ]:
# Create column with first interaction term

X_Shell_weight_origin['Shell_weight_Sex_I_interaction'] = X_Shell_weight_origin['Shell_weight'] * X_Shell_weight_origin['Sex_I']

In [ ]:
# Create column with second interaction term

X_Shell_weight_origin['Shell_Sex_M_interaction'] = X_Shell_weight_origin['Shell_weight'] * X_Shell_weight_origin['Sex_M']

In [ ]:
X_Shell_weight_origin.head()

Shell_weight  Sex_I  Sex_M  Shell_weight_Sex_I_interaction  \
0         0.150  False   True                           0.000   
1         0.070  False   True                           0.000   
2         0.210  False  False                           0.000   
3         0.155  False   True                           0.000   
4         0.055   True  False                           0.055   

   Shell_Sex_M_interaction  
0                    0.150  
1                    0.070  
2                    0.000  
3                    0.155  
4                    0.000

In [ ]:
model_no_interaction = LinearRegression()

In [ ]:
model_no_interaction.fit(X_Shell_weight_origin[['Shell_weight', 'Sex_I', 'Sex_M']], y)

LinearRegression()

In [ ]:
y_pred_no_interaction = model_no_interaction.predict(X_Shell_weight_origin[['Shell_weight', 'Sex_I', 'Sex_M']])

In [ ]:
r2_value_no_inter = r2_score(y, y_pred_no_interaction)

In [ ]:
# Adjusted R squared

1 - (1- r2_value_no_inter) * ( (y.size - 1) / (y.size - 3 - 1) ) # The 3 represents that we have 3 predictors in the model: Shell_weight, dummy 1, and dummy 2

0.4061630202172286

Next, fit the model with interaction

In [ ]:
model_interaction = LinearRegression()

In [ ]:
# When using interaction we use the whole object

model_interaction.fit(X_Shell_weight_origin, y)

LinearRegression()

In [ ]:
y_pred_interaction = model_interaction.predict(X_Shell_weight_origin)

In [ ]:
r2_value_inter = r2_score(y, y_pred_interaction)

In [ ]:
# Adjusted R squared

1 - (1- r2_value_inter) * ( (y.size - 1) / (y.size - 5 - 1) ) # We have 5 predictors (Shell_weight, two dummies, two interactions)

0.42681393447672744

In [ ]:
# % increase
# (interaction - no interaction)/no interaction

(0.42681393447672744 - 0.4061630202172286)/0.4061630202172286

0.050843905603356214

In [ ]:
intercept = model_interaction.intercept_

coefficients = model_interaction.coef_

coef_values = np.concatenate(([intercept], coefficients)) # need to convert the intercept value into a list or array bf using it in concatenate.

column_names = np.concatenate((['Intercept'], model_interaction.feature_names_in_))

# Creating the DataFrame

coefficients_df = pd.DataFrame({'Coefficient Name': column_names, 'Coefficient Value': coef_values})

print(coefficients_df)

                 Coefficient Name  Coefficient Value
0                       Intercept           8.100664
1                    Shell_weight          10.028279
2                           Sex_I          -2.960139
3                           Sex_M          -0.728036
4  Shell_weight_Sex_I_interaction          11.425066
5         Shell_Sex_M_interaction           1.791696


Predicted Rings = β0 + β1 × Shell_weight + β2 × Sex_I + β3 × Sex_M + β4 × (Shell_weight × Sex_I) + β5 × (Shell_weight × Sex_M)

Predicted Rings = 8.1 + 10 * Shell_weight - 2.9 * Sex_I - 0.7 * Sex_M + 11.4 * (Shell_weight * Sex_1) + 1.7 * (Shell_weight * Sex_M)

1) For Female:

Predicted Rings = 8.1 + 10 * Shell_weight

2) For Infant:

Predicted Rings = 8.1 + 10 * Shell_weight − 2.9 + 11.4 * Shell_weight

Predicted Rings = 5.1 + 21.4 * Shell_weight

3) For Male:

Predicted Rings = 8.1 + 10 * Shell_weight − 0.7 + 1.7 * Shell_weight

Predicted Rings = 7.3 + 11.8 × Shell_weight


**Interpretation of interaction**

*   The effect of Shell_weight on Rings is stronger for Infants compared to Females
*   The effect of Shell_weight on Rings is stronger for Males compared to Females.

#3b

In [ ]:
# Define coefficients from the model
intercept_female = 8.100664
coef_shell_weight_female = 10.028279

intercept_male = 7.372628
coef_shell_weight_male = 11.819975

# Define the change in Shell_weight
shell_weight_1 = 0.1
shell_weight_2 = 0.3

# Calculate predicted Rings for Female abalone
predicted_rings_female_1 = intercept_female + coef_shell_weight_female * shell_weight_1
predicted_rings_female_2 = intercept_female + coef_shell_weight_female * shell_weight_2

# Calculate the effect of Shell_weight increase for Female
effect_female = predicted_rings_female_2 - predicted_rings_female_1

# Calculate predicted Rings for Male abalone
predicted_rings_male_1 = intercept_male + coef_shell_weight_male * shell_weight_1
predicted_rings_male_2 = intercept_male + coef_shell_weight_male * shell_weight_2

# Calculate the effect of Shell_weight increase for Male
effect_male = predicted_rings_male_2 - predicted_rings_male_1

# Print results
print(f"Effect on Rings for Females when Shell_weight increases from {shell_weight_1} to {shell_weight_2}: {effect_female:.4f}")
print(f"Effect on Rings for Males when Shell_weight increases from {shell_weight_1} to {shell_weight_2}: {effect_male:.4f}")

Effect on Rings for Females when Shell_weight increases from 0.1 to 0.3: 2.0057
Effect on Rings for Males when Shell_weight increases from 0.1 to 0.3: 2.3640


# Justification
## The effect of increasing Shell_weight from 0.1 to 0.3 is not the same for females and males. It has a stronger impact on Rings for male abalones compared to female abalones, as shown by the larger increase in predicted Rings for males.